### ISSUE TITLE
Better truncation of Japanese posts

### ISSUE THREAD

**Issue by sayunuh** [snarfed/bridgy-fed#1625]
I observe that when a long ActivityPub post has to be truncated for Bluesky, Bridgy Fed does that only at explicit word breaks. This behaviour causes issues handling languages (or writing systems) that don’t use U+0020 SPACE to delimit words or sentences, for example Japanese. Often entire paragraphs are gone.

At least in Japanese you can truncate at basically anywhere in a sentence. It is the same for Chinese. Every Chinese character/hiragana/katakana is a breaking opportunity. I believe you can refer to the Unicode line breaking properties for comprehensiveness.

By the way thanks for the service existing at all, it helps tremendously.

**Comment by snarfed** [comment]
Oh wow, great point, makes sense. Thank you for filing!

Looks like the Unicode line breaking algorithm is http://www.unicode.org/reports/tr14/ . Python has https://pypi.org/project/uniseg/0.6.4/ (native) and https://pypi.org/project/unicode-linebreak/ (Rust wrapped), and also a feature request to add it to `textwrap`, https://github.com/python/cpython/issues/86141.

**Comment by Tamschi** [comment]
Since I implemented Chinese and Japanese word-wrap yesterday (unrelatedly to this project, just a small patch plugin for game development), there's some info on break-excluded characters for CJK languages on pages 85-86 of *Office Open XML Part 4: Markup Language Reference* in the section *2.3.1.16 kinsoku (Use East Asian Typography Rules for First and Last Character per Line)*.

**Comment by sayunuh** [comment]
I thought “just use the existing implementation” would suffice for the purpose of truncation, but since kinsoku is brought up here… To be precise, some punctuation marks prevent line‐breaking in Japanese typography. Opening brackets and the like must not be at the end of a line, and things like commas must not be at the start of a line. 

As for whether they should be taken care of when truncating text, I think end‐of‐line prohibition (gyōmatsu kinsoku) may be obeyed, but start‐of‐line prohibition (gyōtō kinsoku) is better ignored. For example…

- In `これが「不気味の谷」です`, `「不` is a non‐breaking pair because `「` is an end‐of‐line kinsoku character. It can be argued that `これが […]` looks better than `これが「 […]` as an output of truncation, though I can live with both.
- In `おはようございます。`, `す。` is a non‐breaking pair because `。` is a start‐of‐line kinsoku character. But regardless of that fact, I’m confident that `おはようございます […]` is better than `おはようございま […]` as an output of truncation.

So it can be complicated. At the end of the day you may ellipsize anywhere and it will be understandable. Using line‐breaking algorithm will work mostly, but the results might not be ideal since line‐breaking and truncation are slightly different problems.

**Comment by sayunuh** [comment]
Just to be sure — any algorithm that allow truncation at places other than whitespaces is a hundred times more favourable, in my opinion. Where specifically it is allowed doesn’t matter much.

I will leave an example of badly affected posts: https://bsky.app/profile/sayunu.mofu.kemo.no.ap.brid.gy/post/3legn4flqfml2

**Comment by Tamschi** [comment]
I'll mark this as bug to hopefully keep it on the radar a bit more. (It seems I forgot to do that the first time around.)

**Comment by Daft-Freak** [comment]
I had a quick look at this and the truncation happens all the way over in `brevity` https://github.com/kylewm/brevity/blob/master/brevity.py#L181 🤔

**Comment by sayunuh** [comment]
If implementing a “proper” algorithm is not a quick fix, I suggest that in the meantime we just allow truncating mid‐word. Is it worse than an entire paragraph gone? (I’m currently hesitant to suggest people to follow my bridged account because of this behaviour.)

Sorry that I’m not contributing to the code and just complaining.

**Comment by snarfed** [comment]
@sayunuh if I start with the naive dumb approach of allowing truncation anywhere for languages that aren't space-delimited, do you know of a list of the common languages I should include? Maybe https://en.m.wikipedia.org/wiki/Category:Writing_systems_without_word_boundaries ? So ja, zh, th, vi, my, km, lo, ...? And not ko?

**Comment by sayunuh** [comment]
@snarfed Oh, so we use language tags to retain current behaviour for languages like English? I am not particularly knowledgeable beyond that JA and ZH should be included. That Wikipedia list seems alright to me, but as noted in the page, some writing systems use delimiters other than U+0020 SPACE, so making a comprehensible list will not be simple, in its own way.

I will also note that a language‐based approach is not the be‐all, end‐all solution because situations like a Japanese paragraph containing an English word will not be handled in the most ideal way. (Though I personally don’t mind it at all if that word is ellipsized mid‐word.)

**Comment by snarfed** [comment]
Agreed! The "correct" solution here would support text with multiple languages, examine all words and characters, and apply Unicode line breaking character by character.

I'm not ready to tackle that just yet, but I can do the dumb hack you mentioned in https://github.com/snarfed/bridgy-fed/issues/1625#issuecomment-2723608942 . I'd do that based on the language `contentMap` says the post is, and naively assume the whole post is in that language.

**Comment by snarfed** [comment]
Shipped the simple hack. @sayunuh hope it helps!

**Comment by sayunuh** [comment]
@snarfed Did a quick test and it’s working as intended. Thank you for the fix!
